tests/raft/test_raft.rs

```diff
@@ -3429,3 +3451,276 @@ fn test_transfer_non_member() {
         .expect("");;
     assert_eq!(raft.state, StateRole::Follower);
 }
+
+pub fn new_test_learner_raft(
+    id: u64,
+    peers: Vec<u64>,
+    learners: Vec<u64>,
+    election: usize,
+    heartbeat: usize,
+    storage: MemStorage,
+) -> Interface {
+    let mut cfg = new_test_config(id, peers, election, heartbeat);
+    cfg.learners = learners;
+    Interface::new(Raft::new(&cfg, storage))
+}
+
+#[test]
+fn test_learner_election_timeout() {
+    let mut n1 = new_test_learner_raft(1, vec![1], vec![2], 10, 1, new_storage());
+    n1.become_follower(1, INVALID_ID);
+
+    let mut n2 = new_test_learner_raft(2, vec![1], vec![2], 10, 1, new_storage());
+    n2.become_follower(1, INVALID_ID);
+
+    let timeout = n2.get_election_timeout();
+    n2.set_randomized_election_timeout(timeout);
+
+    // n2 is a learner. Learner should not start election even when time out.
+    for _ in 0..timeout {
+        n2.tick();
+    }
+    assert_eq!(n2.state, StateRole::Follower);
+}
+
+#[test]
+fn test_learner_promotion() {
+    let mut n1 = new_test_learner_raft(1, vec![1], vec![2], 10, 1, new_storage());
+    n1.become_follower(1, INVALID_ID);
+
+    let mut n2 = new_test_learner_raft(2, vec![1], vec![2], 10, 1, new_storage());
+    n2.become_follower(1, INVALID_ID);
+
+    let mut network = Network::new(vec![Some(n1), Some(n2)]);
+    assert_eq!(network.peers[&1].state, StateRole::Follower);
+
+    // n1 should become leader.
+    let timeout = network.peers[&1].get_election_timeout();
+    network
+        .peers
+        .get_mut(&1)
+        .unwrap()
+        .set_randomized_election_timeout(timeout);
+    for _ in 0..timeout {
+        network.peers.get_mut(&1).unwrap().tick();
+    }
+    assert_eq!(network.peers[&1].state, StateRole::Leader);
+    assert_eq!(network.peers[&2].state, StateRole::Follower);
+
+    let mut heart_beat = Message::new();
+    heart_beat.set_to(1);
+    heart_beat.set_from(1);
+    heart_beat.set_msg_type(MessageType::MsgBeat);
+    network.send(vec![heart_beat.clone()]);
+
+    // Promote n2 from learner to follower.
+    network.peers.get_mut(&1).unwrap().add_node(2);
+    network.peers.get_mut(&2).unwrap().add_node(2);
+    assert_eq!(network.peers[&2].state, StateRole::Follower);
```

Comment:
Should check if it's learner.

Reply:
In etcd it is
```go
if n2.isLearner {
		t.Error("peer 2 is learner, want not")
}
```
I think the are same because n2 must not be leader?

Reply:
Yes, if there is no bug. Unfortunately you can't assume there is no bug in tests.